// @parcel/transformer-posthtml/lib/PostHTMLTransformer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _plugin = require("@parcel/plugin");

var _posthtml = _interopRequireDefault(require("posthtml"));

var _posthtmlParser = _interopRequireDefault(require("posthtml-parser"));

var _posthtmlRender = _interopRequireDefault(require("posthtml-render"));

var _nullthrows = _interopRequireDefault(require("nullthrows"));

var _semver = _interopRequireDefault(require("semver"));

var _loadPlugins = _interopRequireDefault(require("./loadPlugins"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var _default = new _plugin.Transformer({
  async getConfig({
    asset,
    options
  }) {
    let config = await asset.getConfig(['.posthtmlrc', '.posthtmlrc.js', 'posthtml.config.js'], {
      packageKey: 'posthtml'
    });
    config = config || {}; // load plugins

    config.plugins = await (0, _loadPlugins.default)(config.plugins, asset.filePath, options); // tells posthtml that we have already called parse

    config.skipParse = true;
    return config;
  },

  canReuseAST({
    ast
  }) {
    return ast.type === 'posthtml' && _semver.default.satisfies(ast.version, '^0.4.0');
  },

  async parse({
    asset,
    config
  }) {
    // if we don't have a config it is posthtml is not configure, don't parse
    if (!config) {
      return;
    }

    return {
      type: 'posthtml',
      version: '0.4.1',
      program: (0, _posthtmlParser.default)((await asset.getCode()), {
        lowerCaseAttributeNames: true
      })
    };
  },

  async transform({
    asset,
    config
  }) {
    if (!config) {
      return [asset];
    }

    let ast = (0, _nullthrows.default)((await asset.getAST()));
    let res = await (0, _posthtml.default)(config.plugins).process(ast.program, config);

    if (res.messages) {
      await Promise.all(res.messages.map(({
        type,
        file: filePath
      }) => {
        if (type === 'dependency') {
          return asset.addIncludedFile({
            filePath
          });
        }

        return Promise.resolve();
      }));
    }

    asset.setAST({
      type: 'posthtml',
      version: '0.4.1',
      program: JSON.parse(JSON.stringify(res.tree)) // posthtml adds functions to the AST that are not serializable

    });
    return [asset];
  },

  generate({
    ast
  }) {
    return {
      content: (0, _posthtmlRender.default)(ast.program)
    };
  }

});

exports.default = _default;
